---
title: "IntegrityAlgorithm: documentation"
author: "Audrée Lemieux"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  bookdown::html_document2:
    fig_caption: yes
    toc: true
    number_sections: true
    global_numbering: true
abstract:
  This guide explains the functioning of the R-based algorithm for assessing provirus integrity. Each provirus undergoes analysis to identify the presence of (i) inversions, (ii) hypermutations, (iii) large internal deletions, (iv) stop codons, (v) $\Psi$ defects, and (vi) small internal deletions in *gag*, *pol*, *vif*, *vpr*, *tat*, *rev*, *vpu*, *nef*, *env*, and RRE. Deleterious stop codons in any region except for nef are considered defects. Proviruses devoid of any of these defects are generally regarded as intact. The present algorithm combines the results from LANL's HIV Database QCTools and Gene Cutter, NIH's ProSeq-IT, and manual assessment using Geneious Prime.<br><br>
---

<style type="text/css">
h1.title {
  text-align: center;
}
</style>

```{css echo=FALSE}
.bordered{
  border-style: solid;
}
```

```{css echo=FALSE}
.caption{
  margin: auto;
  text-align: center;
}
```

```{r include=FALSE}
knitr::opts_chunk$set(comment = NA)
```

<!---
-->
<!---
-->

# Resources

**Los Alamos National Laboratory (LANL)**

* HIV Database QCTool: [https://www.hiv.lanl.gov/content/sequence/QC/index.html](https://www.hiv.lanl.gov/content/sequence/QC/index.html)
* HIV Database Gene Cutter: [https://www.hiv.lanl.gov/content/sequence/GENE_CUTTER/cutter.html](https://www.hiv.lanl.gov/content/sequence/GENE_CUTTER/cutter.html)

**NIH National Cancer Institute**

* Proviral Sequence Annotation \& Intactness Test (ProSeq-IT): [https://psd.cancer.gov/tools/pvs_annot.php](https://psd.cancer.gov/tools/pvs_annot.php)

**Related papers**

* Dufour, C., Richard, C., Pardons, M. *et al*. [Phenotypic characterization of single CD4<sup>+</sup> T cells harboring genetically intact and inducible HIV genomes](https://www.nature.com/articles/s41467-023-36772-x). *Nat Commun.* **14**, 1115 (2023).
* Sannier, G., Dubé, M., Dufour, C. *et al*. [Combined single-cell transcriptional, translational, and genomic profiling reveals HIV-1 reservoir diversity](https://www.sciencedirect.com/science/article/pii/S221112472101086X?via%3Dihub). *Cell Rep.* **36**, 109643 (2021).

**Text editors**

For Mac: 

* Default text editor (TextEdit)
* TextMate: [https://macromates.com/](https://macromates.com/)

For Windows: 

* Default text editor (Notepad)
* Notepad++: [https://notepad-plus-plus.org/](https://notepad-plus-plus.org/)

\newpage


<!---
-->
<!---
-->


# Before running the algorithm

Each provirus undergoes analysis to identify the presence of (i) inversions, (ii) hypermutations, (iii) large internal deletions, (iv) stop codons, (v) $\Psi$ defects, and (vi) small internal deletions in *gag*, *pol*, *vif*, *vpr*, *tat*, *rev*, *vpu*, *nef*, *env*, and RRE.

1. <span style="color:#7EB6D6;">**Inversions** (HEX #7EB6D6)</span>: manually identified in the alignment using Geneious
2. <span style="color:#989895;">**Hypermutations** (HEX #989895)</span>: identified with QCTool
3. <span style="color:#CA5F2E;">**Large internal deletions** (HEX #CA5F2E)</span>: size $<$ 8800 bp without primers, can be identified using ProSeq-IT
4. <span style="color:#E6AB48;">**Stop codons** (HEX #E6AB48)</span>: identified with GeneCutter (and QCTool); stop codons in Tat2 and/or Nef are not considered defects
5. <span style="color:#B0B2CB;">**$\Psi$ defects** (HEX #B0B2CB)</span>: manually identified in the alignment using Geneious
    a) Psi deletion
    b) SL2 deletion
    c) MSD point mutation
6. <span style="color:#941100;">**Small internal deletions** (HEX #941100)</span>: identified using ProSeq-IT with the values provided by Sarah Palmer (see [Appendix A, Criteria for defining defects][A. Criteria for defining defects])
    a) $>$  50 nt deleted in *gag*
    b) $>$ 50 nt deleted in *pol*
    c) $>$ 30 nt deleted in *vif*
    d) $>$ 15 nt deleted in *vpr*
    e) $>$ 11 nt deleted in *tat* exon 1 only, or $>$ 15 nt deleted in complete *tat*
    f) $>$ 18 nt deleted in *rev*
    g) $>$ 12 nt deleted in *vpu*
    h) $>$ 31 nt deleted in *nef* (a defective *nef* is not considered a defect for the integrity)
    i) $>$ 100 nt deleted in *env*
7. <span style="color:#76974C;">**Intact** (HEX #76974C)</span>

<!---
-->

## Naming your sequences

Make sure that each sequence name begins with the donor's ID, followed by "_". See example:

```{r, echo = F}
cat(">Donor1_L1p1_C6\n>DONOR2_L1p1_D8_cluster0_ReadCount349_reversed\n>CHO-HIV-72_L1p6_D11\n>DONOR3_L1p1_D9")
```

Avoid these kinds of names, including those containing the reference sequence number:

```{r, echo = F}
cat(">sample_Pacbio_Donor1_L1p1_C6\n>Pacbio82_DONOR2_L1p1_D8_cluster0_ReadCount349_reversed\n>CHO_HIV_72_L1p6_D11\n>DONOR3_L1p1_D9_guide_K03455")
```

**Before** importing your FASTA file into the various tools, double-check the names to (1) remove special characters, and (2) make sure there are no typos (e.g., `DONR1` vs. `Donor 1` vs. `DONOR1`).

Names can be edited with any [text file editor][Resources].

<!---
-->

## HIV Database QCTool

After completing the clean 3, upload your FASTA file containing the aligned sequences without primers and the reference sequence (figure \@ref(fig:Figure1)). It may take up to a day to receive the results via email. **Warning:** the files remain available for up to 4 days.

**Output:** Once the output is available, download the summary `.txt` file (figure \@ref(fig:Figure2)).

<div class="bordered">
```{r Figure1, echo = F, fig.alin = "center", fig.cap = "HIV Database QCTool submission page. Enter your email address for the output to be sent to you once it is available."}
knitr::include_graphics("Images/QCTool.png")
```
</div>

<br>

<div class="bordered">
```{r Figure2, echo = F, fig.alin = "center", fig.cap = "HIV Database QCTool output page. The `Download Summary' button can be found at the end."}
knitr::include_graphics("Images/QCTool_rx.png")
```
</div>

<!---
-->

## HIV Database Gene Cutter

After completing the clean 3, upload your FASTA file containing the _aligned_ sequences without primers and the reference sequence (figure \@ref(fig:Figure3)). This tool is relatively quicker than QCTool. **Warning:** the files remain available for up to 4 days.

<div class="bordered">
```{r Figure3, echo = F, fig.alin = "center", fig.cap = "HIV Gene Cutter submission page. Check (i) the `pre-aligned' and (ii) `email results' boxes."}
knitr::include_graphics("Images/GeneCutter.png")
```
</div>

<!---
-->

## ProSeq-IT

After completing the clean 3, upload your FASTA file containing the aligned sequences without primers and the reference sequence (figure \@ref(fig:Figure4)). The results should be available almost immediately. Make sure that the names do not contain any special characters -- otherwise, the sequence(s) might not appear. These can be edited in any text editor.

**Output:** Once the output is available, download the summary `.xls` file (figure \@ref(fig:Figure5)).

<div class="bordered">
```{r Figure4, echo = F, fig.alin = "center", fig.cap = "ProSeq-IT submission page."}
knitr::include_graphics("Images/Proseq-IT.png")
```
</div>

<br>

<div class="bordered">
```{r Figure5, echo = F, fig.alin = "center", fig.cap = "ProSeq-IT results page. The 'Download Result' button can be found at the beginning."}
knitr::include_graphics("Images/Proseq-IT_rx.png")
```
</div>

<!---
-->

## Preparing the Excel template

Because the inversions and $\Psi$ defects are manually identified with Geneious, an Excel template has to be prepared.

This Excel template contains three tabs:

1. **ProseqIT_criteria (locked):** used for the analysis of ProSeq-IT
2. **Manual_assessment:** contains the results of the manual assessment with Geneious (figure \@ref(fig:Figure6))
3. **Hyperlinks:** contains the hyperlinks sent by email for HIV Database QCTool and Gene Cutter (figure \@ref(fig:Figure7))

<div class="bordered">
```{r Figure6, echo = F, fig.alin = "center", fig.cap = "Excel template, sheet 'Manual_assessment.' Write 'Y' if the sequence contains an inversion and/or a Psi defect, or 'N' otherwise."}
knitr::include_graphics("Images/Manual_assessment.png")
```
</div>

<br>

<div class="bordered">
```{r Figure7, echo = F, fig.alin = "center", fig.cap = "Excel template, sheet 'Hyperlinks.' Copy-paste QCTool's and Gene Cutter's output hyperlinks in the appropriate cells."}
knitr::include_graphics("Images/Hyperlinks.png")
```
</div>

\newpage

<!---
-->
<!---
-->

# Running the algorithm

## About the algorithm

For analysing the intactness of HIV proviruses, the main function is called `HIV_IntegrityAnalysis()`.

1. The algorithm will first analyse the results from QCTool to identify ORFs containing stop codons and hypermutated sequences.
2. It will then analyse the results from Gene Cutter to obtain the start and stop codons, if applicable, of and within each ORF.
3. Next, it will analyse the results from ProSeq-IT for large and small internal deletions.
4. Finally, all results from QCTool, Gene Cutter, ProSeq-IT, and Geneious (through the Excel template) will be combined to assess the intactness of the HIV proviruses.

```{r HIV_IA1, eval = F}
HIV_IntegrityAnalysis(template_filename, QCTool_summary, ProseqIT_rx, ProseqIT_RefSeq, RefSeq)
# Description: analyses the intactness of HIV proviruses
# Arguments:
#   - template_filename: the name of the Excel template file
#   - QCTool_summary: the name of the summary .txt file downloaded from QCTool's output
#   - ProseqIT_rx: the name of the summary .xls file downloaded from ProSeq-IT
#   - ProseqIT_RefSeq: If TRUE, the reference sequence is included in ProSeq-IT's results. Default is FALSE
#   - RefSeq: If TRUE, the reference sequence is included in QCTool's and Gene Cutter's results. Default is TRUE
# Functions called: QCTool_analyses, GeneCutter_analyses, ProseqIT_analyses, IntegrateInfo
```

<!---
-->

## Downloading and loading R packages

The algorithm requires multiple R packages (`readxl`, `RCurl`, `XML`, and `stringr`). To download them (only has to be done once) on your computer, run the command "`install.packages()`".

```{r install_libraries, eval = F}
install.packages("readxl")
install.packages("RCurl")
install.packages("XML")
install.packages("stringr")
```

Once they are downloaded, run the command "`library()`" **every time** you run the script.

```{r load_libraries, eval = F}
library(readxl)
library(RCurl)
library(XML)
library(stringr)
```

<!---
-->

## The four core functions

```{r core_fctns, eval = F}
QCTool_analyses(filename, summary, RefSeq)
# Description: analyses the results from QCTool
# Arguments:
#   - filename: the name of the Excel template
#   - summary: the name of the summary .txt file downloaded from QCTool's output
#   - RefSeq: If TRUE, the reference sequence is included in QCTool's and Gene Cutter's results. Default is TRUE

GeneCutter_analyses(filename, RefSeq)
# Description: analyses the results from Gene Cutter
# Arguments:
#   - filename: the name of the Excel template
#   - RefSeq: If TRUE, the reference sequence is included in QCTool's and Gene Cutter's results. Default is TRUE
# Functions called: parse_GeneCutter, parse_GeneCutter_Stop

ProseqIT_analyses(filename, ProseqIT_filename, ProseqIT_RefSeq)
# Description: analyses the results from ProSeq-IT
# Arguments:
#   - filename: the name of the Excel template
#   - ProseqIT_filename: the name of the summary .xls file downloaded from ProSeq-IT
#   - ProseqIT_RefSeq: If TRUE, the reference sequence is included in ProSeq-IT's results. Default is FALSE
# Functions called: add_to_comment, gag_small_delet, pol_small_delet, vif_small_delet, vpr_small_delet,
#                   tat_small_delet, rev_small_delet, vpu_small_delet, env_small_delet, nef_small_delet

IntegrateInfo()
# Description: integrates the analysed results from QCTool, Gene Cutter, Proseq-IT, and Geneious (manual
#              assessment)
# Functions called: add_info_per_donor, add_info_all
```

<!---
-->

## Running the algorithm

```{r, include = F}
source("Script/IntegrityAlgorithm_clean.R", local = knitr::knit_global())
```

**Before** running the algorithm, make sure that these files are within the same directory (folder) than the `IntegrityAlgorithm.R` script:

1. Excel template
2. QCTool summary `.txt` file
3. ProSeq-IT `.xls` result file

If any file has to be manually edited (e.g., typo in QCTool/Gene Cutter/ProSeq-IT output), **do not** edit with Excel. Use any text editor instead.

```
! Use the `tab` key to autocomplete !
```

### Arguments

To main function `HIV_IntegrityAnalysis()` has 5 arguments; 3 required and 2 optional.

For example:

**Required arguments - quotation marks required**

1. template_filename: `Template_Integrity.xlsx`
2. QCTool_summary: `QCTool_results.txt`
3. ProseqIT_rx: `1901771461.annot.xls`

**Optional arguments - no quotation marks, capitalization required**

1. ProseqIT_RefSeq: in some cases, ProSeq-IT shows the reference sequence. If this is the case, write `TRUE` to make sure that it is not included in the algorithm. If this is not the case, simply do not write anything (default: FALSE) or write `FALSE`.
2. RefSeq: the reference sequence is usually shown in QCTool and Gene Cutter. If this is the case, simply do not write anything (default: TRUE) or write `TRUE` to make sure that it is not included in the algorithm. If this is not the case, write `FALSE`.

### Command

Select all the lines from the `IntegrityAlgorithm.R` script and hit `Run`, `[CTRL] + [Enter]` (Windows), or `[CMD] + [ENTER]` (Mac).

**Default**

The reference sequence is not contained in ProSeq-IT's output, but is in QCTool's and Gene Cutter's:

```{r ex1}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrity.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

**Alternative - the reference sequence is in all three tools outputs**

```{r ex2, eval = F}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrity.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls", ProseqIT_RefSeq = TRUE)
```

<!---
-->

## Outputs

CSV files of the individual analyzes (QCTool, Gene Cutter, and ProSeq-IT) can be found in the 'tmp' folder.

The `intactness_detailedsummary.csv` file contains **all** the defects for all sequences, meaning that defects are not hierarchized. 

The final output can be found in the 'FINAL OUTPUT' folder. It contains the list of hierarchized defects for all sequences (figure \@ref(fig:Figure8)).

<div class="bordered">
```{r Figure8, echo = F, fig.alin = "center", fig.alin = "center", fig.cap = "CSV file containing a summary of hierarchized defects."}
knitr::include_graphics("Images/Defects_summary.png")
```
</div>

\newpage

<!---
-->
<!---
-->

# Troubleshooting

## Gene Cutter error

This error seems to occur with Gene Cutter when inversions are not corrected in the FASTA file submitted (figure \@ref(fig:Figure9)).

<div class="bordered">
```{r Figure9, echo = F, fig.alin = "center", fig.alin = "center", fig.cap = "Error occurring with Gene Cutter."}
knitr::include_graphics("Images/GeneCutter_error.png")
```
</div>

## No such file or directory

This error specifically happens when R cannot create a folder `tmp` or `FINAL_OUTPUT`, usually when the user is not the administrator. Try manually creating the 'tmp' and 'FINAL_OUTPUT' folders and then, run the command again.

```{r, include = F}
source("Script/IntegrityAlgorithm_clean_tmp.R", local = knitr::knit_global())
```

```{r, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrity.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

## 'path' does not exist

The file mentioned is not in the same folder as the R script, or there is a typo.

```{r, include = F}
source("Script/IntegrityAlgorithm_clean.R", local = knitr::knit_global())
```

```{r ts1, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrit.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

## Sheet not found

The `.xlsx` specified is not the right Excel template file, as it is at least missing the sheet 'Hyperlinks.'

```{r ts2, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "AIM panels.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

## Error before analyzing QCTool and/or Gene Cutter

Instead, the error "`subscript out of bounds`" appears. This indicates that the hyperlinks for either QCTool and/or Gene Cutter are not available anymore or incorrect (typos).

```{r ts3, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrity_QC.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

```{r ts4, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis(template_filename = "Template_Integrity_GC.xlsx", QCTool_summary = "QCTool_results.txt", ProseqIT_rx = "1901771461.annot.xls")
```

## Arguments imply different number of rows (when analyzing ProSeq-IT)

This error occurs when ProSeq-IT's output contains the reference sequence and the argument `ProseqIT_RefSeq = TRUE` has not been specified.

```{r ts5, error = T}
library(readxl)
library(RCurl)
library(XML)
library(stringr)

HIV_IntegrityAnalysis("ST121b Inversions Psi Template.xlsx", "summary.txt", "1386232661.annot.xls")
```

\newpage

# Appendix

## A. Criteria for defining defects

Sequences $>$ 8800 bp (full-length) should be run through the PSD with cross referencing to gene cutter to determine intactness

Below is a summary of the output columns from PSD ([https://psd.cancer.gov/](https://psd.cancer.gov/)) with additional criteria where relevant 

Note: for accessory proteins, PSD only specifies the defect, it does not have criteria for intactness. We require all accessory proteins to be intact with the exception of tat (only exon 1 intact (1)) and nef. Criteria for the tolerance of deletions are included in the table below based on deletion of >5% of length of ORF (but keep in mind participant specific deletions). 

If unsure about a sequence – present to group for consensus decision and note in table for future

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-1i2r{background-color:#f8ff00;border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-zv36{background-color:#ffffff;border-color:inherit;font-weight:bold;text-align:left;vertical-align:top}
.tg .tg-fymr{border-color:inherit;font-weight:bold;text-align:left;vertical-align:top}
.tg .tg-c6of{background-color:#ffffff;border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
.tg .tg-y698{background-color:#efefef;border-color:inherit;text-align:left;vertical-align:top}
</style>
<table class="tg">
<thead>
  <tr>
    <th class="tg-zv36">Criteria</th>
    <th class="tg-zv36">PSD comment</th>
    <th class="tg-zv36">HXB2 position</th>
    <th class="tg-zv36">Present in FLIPS sequences</th>
    <th class="tg-zv36">Critical for intactness</th>
    <th class="tg-zv36">Comment</th>
    <th class="tg-fymr">Additional criteria</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-c6of">5LTR_length</td>
    <td class="tg-c6of">5’ LTR length</td>
    <td class="tg-c6of">1-634</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">TAR_stem_loop_insertion</td>
    <td class="tg-c6of">Insertions in TAR stem loop</td>
    <td class="tg-c6of">453-513</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">TAR_stem_loop_deletion</td>
    <td class="tg-c6of">Deletions in TAR stem loop</td>
    <td class="tg-c6of">453-513</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">polyA_insertion</td>
    <td class="tg-c6of">insertions in PolyA region</td>
    <td class="tg-c6of">527-532</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">polyA_deletion</td>
    <td class="tg-c6of">Deletions in PolyA region</td>
    <td class="tg-c6of">527-532</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">u5_gag_pair_R1_insertion</td>
    <td class="tg-c6of">Insertions in the region of U5 that pairs with gag start AUG</td>
    <td class="tg-c6of">559-569</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">u5_gag_pair_R1_deletion</td>
    <td class="tg-c6of">Deletions in the first region of U5 that pairs with gag start AUG</td>
    <td class="tg-c6of">559-569</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">half_pbs_insertion</td>
    <td class="tg-c6of">Insertions in the first 8 bases of PBS</td>
    <td class="tg-c6of">636-644</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">half_pbs_deletion</td>
    <td class="tg-c6of">Deletions in the first 8 bases of PBS</td>
    <td class="tg-c6of">636-644</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS sequences</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">MSD_status</td>
    <td class="tg-1i2r">Mutations or deletions in major splice donor site</td>
    <td class="tg-1i2r">744-745</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: present; no deletion; no mutation</td>
    <td class="tg-0pky">Mutant /deleted MSD can be salvaged with intact cryptic donor site (GT at position 748-749) (2, 3)</td>
  </tr>
  <tr>
    <td class="tg-c6of">package_insertion</td>
    <td class="tg-c6of">Insertions in the packaging signal region</td>
    <td class="tg-c6of">695-810</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">package_deletion</td>
    <td class="tg-1i2r">Deletions in the packaging signal region</td>
    <td class="tg-1i2r">695-810</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: deletions &lt; 8 nucleotides. (Asked Wei Shao why 8 nt? Said on advice from Wu-Shau Hu)</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">U5_gag_pair_R2_insertion</td>
    <td class="tg-c6of">Insertions in gag start AUG region that pairs with U5</td>
    <td class="tg-c6of">788-798</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">U5_gag_pair_R2_deletion</td>
    <td class="tg-1i2r">Deletions in gag start AUG region that pairs with U5</td>
    <td class="tg-1i2r">788-798</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: deletion &lt; 3 nucleotides. Region required for dimerization (4)</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">2base_before_gag_status</td>
    <td class="tg-1i2r">Deletions in the 2 bases before gag AUG</td>
    <td class="tg-1i2r">788-789</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no deletion. Prevents frameshifts of gag start codon</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">gag_start_codon</td>
    <td class="tg-1i2r">Start codon present</td>
    <td class="tg-1i2r">790-792</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no deletions; no mutations</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">gag_insertion</td>
    <td class="tg-1i2r">Insertions in gag</td>
    <td class="tg-1i2r">790-2292</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;50 nt insertion</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">gag_deletion</td>
    <td class="tg-1i2r">Deletions in gag</td>
    <td class="tg-1i2r">790-2292</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;50 nt deletion</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-y698">gag_frameshift</td>
    <td class="tg-y698">Frameshifts in gag</td>
    <td class="tg-y698">790-2292</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshifts</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">gag_stop_codon</td>
    <td class="tg-1i2r">Out of frame stop codons in gag</td>
    <td class="tg-1i2r">790-2292</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-1i2r">pol_insertion</td>
    <td class="tg-1i2r">Insertions in pol</td>
    <td class="tg-1i2r">2085-5096</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;50 nt insertion</td>
    <td class="tg-0pky" rowspan="4">Pol requires and intact gag start codon for transcription</td>
  </tr>
  <tr>
    <td class="tg-1i2r">pol_deletion</td>
    <td class="tg-1i2r">Deletions in pol</td>
    <td class="tg-1i2r">2085-5096</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;50 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">pol_frameshift</td>
    <td class="tg-y698">Frameshifts in pol</td>
    <td class="tg-y698">2085-5096</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">pol_stop_codon</td>
    <td class="tg-1i2r">out of frame stop codons in pol</td>
    <td class="tg-1i2r">2085-5096</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-c6of">vif_insertion</td>
    <td class="tg-c6of">Insertion in vif</td>
    <td class="tg-c6of">5041-5619</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check vif start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vif_deletion</td>
    <td class="tg-1i2r">Deletion in vif</td>
    <td class="tg-1i2r">5041-5619</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;30 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">vif_frameshift</td>
    <td class="tg-y698">Frameshift in vif</td>
    <td class="tg-y698">5041-5619</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vif_stop_codon</td>
    <td class="tg-1i2r">Stop codons in vif</td>
    <td class="tg-1i2r">5041-5619</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-c6of">vpr_insertion</td>
    <td class="tg-c6of">Insertion in vpr</td>
    <td class="tg-c6of">5559-5850</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check vpr start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vpr_deletion</td>
    <td class="tg-1i2r">Deletion in vpr</td>
    <td class="tg-1i2r">5559-5850</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;15 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">vpr_frameshift</td>
    <td class="tg-y698">Frameshift in vpr</td>
    <td class="tg-y698">5559-5850</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vpr_stop_codon</td>
    <td class="tg-1i2r">Stop codons in vpr</td>
    <td class="tg-1i2r">5559-5850</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-c6of">tat_insertion</td>
    <td class="tg-c6of">Insertion in tat</td>
    <td class="tg-c6of">Exon 1: 5831-6045; Exon 2: 8370-8469</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check tat start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">tat_deletion</td>
    <td class="tg-1i2r">Deletion in tat</td>
    <td class="tg-1i2r">Exon 1: 5831-6045; Exon 2: 8370-8469</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;11 nt in exon 1 ; &lt;15 nt overall sequence</td>
  </tr>
  <tr>
    <td class="tg-y698">tat_frameshift</td>
    <td class="tg-y698">Frameshift in tat</td>
    <td class="tg-y698">Exon 1: 5831-6045; Exon 2: 8370-8469</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift in exon 1</td>
  </tr>
  <tr>
    <td class="tg-1i2r">tat_stop_codon</td>
    <td class="tg-1i2r">Stat codons in tat</td>
    <td class="tg-1i2r">Exon 1: 5831-6045; Exon 2: 8370-8469</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons in exon 1</td>
  </tr>
  <tr>
    <td class="tg-c6of">rev_insertion</td>
    <td class="tg-c6of">Insertion in rev</td>
    <td class="tg-c6of">Exon 1: 5970-6045; Exon 2: 8379-8653</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check rev start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">rev_deletion</td>
    <td class="tg-1i2r">Deletion in rev</td>
    <td class="tg-1i2r">Exon 1: 5970-6045; Exon 2: 8379-8653</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;18 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">rev_frameshift</td>
    <td class="tg-y698">Frameshift in rev</td>
    <td class="tg-y698">Exon 1: 5970-6045; Exon 2: 8379-8653</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">rev_stop_codon</td>
    <td class="tg-1i2r">Stop codons in rev</td>
    <td class="tg-1i2r">Exon 1: 5970-6045; Exon 2: 8379-8653</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-c6of">vpu_insertion</td>
    <td class="tg-c6of">Insertion in vpu</td>
    <td class="tg-c6of">6062-6310</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check vpu start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vpu_deletion</td>
    <td class="tg-1i2r">Deletion in vpu</td>
    <td class="tg-1i2r">6062-6310</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt;12 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">vpu_frameshift</td>
    <td class="tg-y698">Frameshift in vpu</td>
    <td class="tg-y698">6062-6310</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">vpu_stop_codon</td>
    <td class="tg-1i2r">Stop codons in vpu</td>
    <td class="tg-1i2r">6062-6310</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons<br><span style="font-weight:400;font-style:normal">*note that it is common to have a stop codon 5 nt from end of vpu (W to stop); but this W residue is important for virion release from cell surface (5)</span><br></td>
  </tr>
  <tr>
    <td class="tg-c6of">nef_insertion</td>
    <td class="tg-c6of">Insertion in nef</td>
    <td class="tg-c6of">8797-9417</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky" rowspan="4">Check nef start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">nef_deletion</td>
    <td class="tg-1i2r">Deletion in nef</td>
    <td class="tg-1i2r">8797-9417</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">No</td>
    <td class="tg-1i2r">Intact: &lt;31 nt deletion</td>
  </tr>
  <tr>
    <td class="tg-y698">nef_frameshift</td>
    <td class="tg-y698">Frameshift in nef</td>
    <td class="tg-y698">8797-9417</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">No</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">nef_stop_codon</td>
    <td class="tg-1i2r">Stop codons in nef</td>
    <td class="tg-1i2r">8797-9417</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">No</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-1i2r">env_insertion</td>
    <td class="tg-1i2r">Insertion in env</td>
    <td class="tg-1i2r">6225-8795</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt; 150 nucleotides</td>
    <td class="tg-0pky" rowspan="4">Check env start codon intact</td>
  </tr>
  <tr>
    <td class="tg-1i2r">env_deletion</td>
    <td class="tg-1i2r">Deletion in env</td>
    <td class="tg-1i2r">6225-8795</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &lt; 100 nucleotide</td>
  </tr>
  <tr>
    <td class="tg-y698">env_frameshift</td>
    <td class="tg-y698">Frameshift in env</td>
    <td class="tg-y698">6225-8795</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Yes</td>
    <td class="tg-y698">Intact: no frameshift</td>
  </tr>
  <tr>
    <td class="tg-1i2r">env_stop_codon</td>
    <td class="tg-1i2r">Stop codons in env</td>
    <td class="tg-1i2r">6225-8795</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: no premature stop codons</td>
  </tr>
  <tr>
    <td class="tg-1i2r">rre_status</td>
    <td class="tg-1i2r">deletions in RRE region</td>
    <td class="tg-1i2r">7710-8061</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Yes</td>
    <td class="tg-1i2r">Intact: &gt; 233 nucleotides in length and key domain deletion &lt;10 nucleotides</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">3LTR_length</td>
    <td class="tg-c6of">3’ LTR length</td>
    <td class="tg-c6of">9086-9719</td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of"></td>
    <td class="tg-c6of">Irrelevant for FLIPS</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">two_LTR_similarity</td>
    <td class="tg-c6of">The similarity between 5’ and 3’ LTR</td>
    <td class="tg-c6of"></td>
    <td class="tg-c6of">Yes</td>
    <td class="tg-c6of">No</td>
    <td class="tg-c6of">Irrelevant for FLIPS</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-c6of">inferred_intactness</td>
    <td class="tg-c6of">Whether the sequences are intact</td>
    <td class="tg-c6of"></td>
    <td class="tg-c6of"></td>
    <td class="tg-c6of"></td>
    <td class="tg-c6of"></td>
    <td class="tg-0pky"></td>
  </tr>
</tbody>
</table>

<br>

**References**

1. Clark E, Nava B, Caputi M. Tat is a multifunctional viral protein that modulates cellular gene expression and functions. Oncotarget. 2017;8(16):27569-81.
2. Pinzone MR, VanBelzen DJ, Weissman S, Bertuccio MP, Cannon L, Venanzi-Rullo E, et al. Longitudinal HIV sequencing reveals reservoir expression leading to decay which is obscured by clonal expansion. Nat Commun. 2019;10(1):728.
3. Purcell DF, Martin MA. Alternative splicing of human immunodeficiency virus type 1 mRNA modulates viral protein expression, replication, and infectivity. J Virol. 1993;67(11):6365-78.
4. Lu K, Heng X, Garyu L, Monti S, Garcia EL, Kharytonchyk S, et al. NMR detection of structures in the HIV-1 5'-leader RNA that regulate genome packaging. Science (New York, NY). 2011;334(6053):242-5.
5. González ME. Vpu Protein: The Viroporin Encoded by HIV-1. Viruses. 2015;7(8):4352-68.

<br>

